File: leetcode/24.swap-nodes-in-pairs/cpp/swappairs.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    // N, 1
    ListNode *swapPairs(ListNode *head)
    {
        ListNode dummyInstance{0, head};
        auto dummy = &dummyInstance;

        auto prev = dummy;
        auto curr = head;

        while (curr != nullptr && curr->next != nullptr)
        {
            const auto next = curr->next;
            const auto nNext = next->next;

            prev->next = next;
            curr->next = nNext;
            next->next = curr;

            prev = curr;
            curr = nNext;
        }

        return dummy->next;
    }
};

File: leetcode/24.swap-nodes-in-pairs/go/swappairs.go
package swappairs

type ListNode struct {
	Val  int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	dummy := ListNode{}
	dummy.Next = head

	prev := &dummy
	curr := dummy.Next

	for curr != nil && curr.Next != nil {
		next := curr.Next
		nNext := next.Next

		prev.Next = next
		next.Next = curr
		curr.Next = nNext

		prev = curr
		curr = nNext
	}

	return dummy.Next
}
